File: packages/neft-input/native/android/TextInputItem.java
package io.neft.extensions.input_extension;

import android.content.Context;
import android.graphics.Color;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.util.Log;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import io.neft.renderer.NativeItem;
import io.neft.renderer.annotation.OnCall;
import io.neft.renderer.annotation.OnCreate;
import io.neft.renderer.annotation.OnSet;
import io.neft.utils.ColorValue;

import java.util.HashMap;
import java.util.Map;

public class TextInputItem extends NativeItem {
    private static final Map<String, Integer> KEYBOARD_TYPES = new HashMap<>();
    private static final Map<String, Integer> KEY_TYPES = new HashMap<>();

    static {
        KEYBOARD_TYPES.put("text", InputType.TYPE_CLASS_TEXT);
        KEYBOARD_TYPES.put("numeric", InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL);
        KEYBOARD_TYPES.put("email", InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS);
        KEYBOARD_TYPES.put("tel", InputType.TYPE_CLASS_PHONE);

        KEY_TYPES.put("done", EditorInfo.IME_ACTION_DONE);
        KEY_TYPES.put("go", EditorInfo.IME_ACTION_GO);
        KEY_TYPES.put("next", EditorInfo.IME_ACTION_NEXT);
        KEY_TYPES.put("search", EditorInfo.IME_ACTION_SEARCH);
        KEY_TYPES.put("send", EditorInfo.IME_ACTION_SEND);
        KEY_TYPES.put(null, EditorInfo.IME_ACTION_UNSPECIFIED);
    }

    int keyboardTypeMod;
    int multilineMod;
    boolean secureTextEntry;

    @OnCreate("TextInput")
    public TextInputItem() {
        super(new EditText(APP.getActivity().getApplicationContext()));
        EditText itemView = getItemView();
        itemView.setBackground(null);
        itemView.setPadding(0, 0, 0, 0);

        setTextColor(new ColorValue(Color.BLACK));

        itemView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                pushEvent("textChange", getItemView().getText().toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });

        setKeyboardType("text");
        setMultiline(false);
        setReturnKeyType(null);
        setSecureTextEntry(false);
    }

    private EditText getItemView() {
        return (EditText) itemView;
    }

    private void updateInputType() {
        int secureTextEntryMod = 0;
        if (secureTextEntry) {
            if ((keyboardTypeMod & InputType.TYPE_CLASS_NUMBER) > 0) {
                secureTextEntryMod = InputType.TYPE_NUMBER_VARIATION_PASSWORD;
            } else {
                secureTextEntryMod = InputType.TYPE_TEXT_VARIATION_PASSWORD;
            }
        }
        getItemView().setInputType(keyboardTypeMod | multilineMod | secureTextEntryMod);
    }

    @OnSet("text")
    public void setText(String val) {
        getItemView().setText(val);
        updateSize();
    }

    @OnSet("textColor")
    public void setTextColor(ColorValue val) {
        getItemView().setTextColor(val.getColor());
    }

    @OnSet("placeholder")
    public void setPlaceholder(String val) {
        getItemView().setHint(val);
    }

    @OnSet("placeholderColor")
    public void setPlaceholderColor(ColorValue val) {
        getItemView().setHintTextColor(val.getColor());
    }

    @OnSet("keyboardType")
    public void setKeyboardType(String keyboardType) {
        Integer val = KEYBOARD_TYPES.get(keyboardType);
        if (val == null) {
            Log.w("Neft", "Unknown TextInput keyboardType " + keyboardType);
            val = InputType.TYPE_CLASS_TEXT;
        }
        keyboardTypeMod = val;
        updateInputType();
    }

    @OnSet("multiline")
    public void setMultiline(boolean multiline) {
        multilineMod = multiline ? InputType.TYPE_TEXT_FLAG_MULTI_LINE : 0;
        updateInputType();
    }

    @OnSet("returnKeyType")
    public void setReturnKeyType(String keyType) {
        Integer val = KEY_TYPES.get(keyType);
        if (val == null) {
            Log.w("Neft", "Unknown TextInput returnKeyType " + keyType);
            val = EditorInfo.IME_ACTION_UNSPECIFIED;
        }
        getItemView().setImeOptions(val);
    }

    @OnSet("secureTextEntry")
    public void setSecureTextEntry(boolean val) {
        secureTextEntry = val;
        updateInputType();
    }

    @OnCall("focus")
    public void focus() {
        EditText itemView = getItemView();
        itemView.requestFocusFromTouch();
        InputMethodManager manager = (InputMethodManager) APP.getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        manager.showSoftInput(itemView, 0);
    }
}

File: packages/ui/neft-input/native/ios/TextInputItem.swift
import UIKit

extension Extension.Input {
    class TextInputItem: NativeItem {
        override static func main() {
            NativeItemBinding()
                .onCreate("TextInput") { TextInputItem() }
                .onSet("text") { (item, val: String) in item.setText(val) }
                .onSet("textColor") { (item, val: UIColor?) in item.setTextColor(val) }
                .onSet("placeholder") { (item, val: String) in item.setPlaceholder(val) }
                .onSet("placeholderColor") { (item, val: UIColor?) in item.setPlaceholderColor(val) }
                .onSet("keyboardType") { (item, val: String) in item.setKeyboardType(val) }
                .onSet("multiline") { (item, val: Bool) in item.setMultiline(val) }
                .onSet("returnKeyType") { (item, val: String) in item.setReturnKeyType(val) }
                .onSet("secureTextEntry") { (item, val: Bool) in item.setSecureTextEntry(val) }
                .onCall("focus") { item in item.focus() }
                .finalize()
        }

        private static let keyboardTypes: [String: UIKeyboardType] = [
            "text": .default,
            "numeric": .decimalPad,
            "email": .emailAddress,
            "tel": .phonePad
        ]

        private static let returnKeyTypes: [String: UIReturnKeyType] = [
            "done": UIReturnKeyType.done,
            "go": UIReturnKeyType.go,
            "next": UIReturnKeyType.next,
            "search": UIReturnKeyType.search,
            "send": UIReturnKeyType.send
        ]

        private var fieldView = UITextField()

        private var placeholder: String = "" {
            didSet {
                updatePlaceholder()
            }
        }

        private var placeholderColor: UIColor? {
            didSet {
                updatePlaceholder()
            }
        }

        init() {
            super.init(itemView: fieldView)
            fieldView.addTarget(self, action: #selector(onTextChange(textField:)), for: UIControl.Event.editingChanged)
            fieldView.addTarget(self, action: #selector(onExit(textField:)), for: UIControl.Event.editingDidEndOnExit)
            fieldView.borderStyle = .none
            fieldView.frame.size.width = 250
            fieldView.frame.size.height = 30
            fieldView.frame = fieldView.frame
            setTextColor(UIColor.black)

            NotificationCenter.default.addObserver(self, selector: #selector(didBeginEditing(_:)), name: UITextField.textDidBeginEditingNotification, object: fieldView)

            NotificationCenter.default.addObserver(self, selector: #selector(didEndEditing(_:)), name: UITextField.textDidEndEditingNotification, object: fieldView)
        }

        func setText(_ val: String) {
            fieldView.text = val
        }

        func setTextColor(_ val: UIColor?) {
            fieldView.textColor = val
        }

        func setPlaceholder(_ val: String) {
            placeholder = val
        }

        func setPlaceholderColor(_ val: UIColor?) {
            placeholderColor = val
        }

        func setKeyboardType(_ val: String) {
            fieldView.keyboardType = TextInputItem.keyboardTypes[val] ?? UIKeyboardType.default
        }

        func setMultiline(_ val: Bool) {
            // TODO
        }

        func setReturnKeyType(_ val: String) {
            fieldView.returnKeyType = TextInputItem.returnKeyTypes[val] ?? UIReturnKeyType.default
        }

        func setSecureTextEntry(_ val: Bool) {
            fieldView.isSecureTextEntry = val
        }

        func focus() {
            fieldView.becomeFirstResponder()
        }

        private func updatePlaceholder() {
            if placeholderColor == nil {
                fieldView.placeholder = placeholder
            } else {
                let attrs = NSAttributedString(string: placeholder,
                                               attributes: [NSAttributedString.Key.foregroundColor: placeholderColor!])
                fieldView.attributedPlaceholder = attrs
            }
        }

        @objc
        private func onTextChange(textField: UITextField) {
            pushEvent(event: "textChange", args: [textField.text ?? ""])
        }

        @objc
        private func onExit(textField: UITextField) {
            App.app.client.pushAction(OutAction.keyRelease, "Enter")
        }

        @objc
        private func didBeginEditing(_ notification: Notification) {
            focused = true
        }

        @objc
        private func didEndEditing(_ notification: Notification) {
            focused = false
        }
    }
}
